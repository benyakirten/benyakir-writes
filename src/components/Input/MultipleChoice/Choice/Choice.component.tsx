import * as React from "react";

import { ChoiceProps } from "@Types/props";
import { ChoiceContainer } from "./Choice.styles";

import { titleToKebab } from "@Utils/strings";

const Choice: React.FC<ChoiceProps> = ({ label, onSelect, value, tabIndex = 0 }) => {
    const _onSelect = React.useCallback(() => onSelect(label), [onSelect, label]);
    return (
        <ChoiceContainer
            role="checkbox"
            aria-checked={value}
            onClick={() => tabIndex >= 0 && _onSelect()}
            checked={value}
            tabIndex={tabIndex}
            hidden={tabIndex < 0}
            aria-labelledby={`${label}-label`}
            data-cy={`select-${titleToKebab(label)}`}
        >
            <span style={{ height: tabIndex < 0 ? 0 : 'auto' }} id={`${label}-label`}>{label}</span>
        </ChoiceContainer>
    );
};

export default Choice;
