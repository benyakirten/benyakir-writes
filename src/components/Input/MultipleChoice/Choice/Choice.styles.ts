import styled from 'styled-components'

import { media } from '@Styles/queries'
import { FONT_LG, TRANSITION_SLOW, Z_BASE, Z_UNDER } from '@StyleVars'

export const ChoiceContainer = styled.div<{
  checked: boolean
  hidden: boolean
}>`
  position: relative;
  z-index: ${(props) => (props.hidden ? Z_UNDER : Z_BASE)};

  cursor: ${(props) => (props.hidden ? 'default' : 'pointer')};

  display: flex;
  justify-content: center;
  align-items: center;

  padding: ${(props) => (props.hidden ? 0 : '1rem')};
  margin: 1rem;
  margin-left: 0;

  height: ${(props) => (props.hidden ? 0 : 'auto')};
  width: 24%;

  ${media.desktop} {
    width: 45%;
  }

  ${media.phone} {
    width: 75%;
  }

  border-radius: 2px;

  background-color: ${(props) =>
    props.checked
      ? props.theme.multipleChoice.textColor
      : props.theme.multipleChoice.background};
  color: ${(props) =>
    props.checked
      ? props.theme.multipleChoice.background
      : props.theme.multipleChoice.textColor};

  font-size: ${FONT_LG};

  transition: all ${TRANSITION_SLOW};
`
