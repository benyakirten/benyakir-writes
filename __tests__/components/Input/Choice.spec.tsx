import * as React from "react";
import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import Choice from "@Input/MultipleChoice/Choice/Choice.component"

describe('Choice component', () => {
    const selectSpy = jest.fn()

    beforeEach(() => {
        selectSpy.mockClear()
    })

    afterEach(cleanup)

    it('should render correctly', () => {
        expect(() => render(<Choice label="test label" value={false} onSelect={selectSpy} />)).not.toThrow()
    })

    it('should render a div with a role of checkbox that\'s labelled by a span', async () => {
        render(<Choice label="test label" value={false} onSelect={selectSpy} />)
        const checkbox = await screen.findByRole("checkbox")
        expect(checkbox.getAttribute("aria-labelledby")).toEqual("test label-label")
        const label = checkbox.firstElementChild!
        expect(label.id).toEqual("test label-label")
    })

    it('should change the background and text color based on the value prop', () => {
        const comp = renderer.create(<Choice label="test label" value={false} onSelect={selectSpy} />)
        const snapOne = comp.toJSON()
        expect(snapOne).toMatchSnapshot()
        expect(snapOne).toHaveStyleRule("background-color", "#343a40")
        expect(snapOne).toHaveStyleRule("color", "#FDF3D8")
        
        comp.update(<Choice label="test label" value={true} onSelect={selectSpy} />)
        const snapTwo = comp.toJSON()
        expect(snapTwo).toMatchSnapshot()
        expect(snapTwo).toHaveStyleRule("background-color", "#FDF3D8")
        expect(snapTwo).toHaveStyleRule("color", "#343a40")
    })

    it('should emit the onSelect function when it is clicked', async () => {
        render(<Choice label="test label" value={false} onSelect={selectSpy} />)
        const checkbox = await screen.findByRole("checkbox")
        fireEvent.click(checkbox)
        expect(selectSpy).toHaveBeenCalledTimes(1)
        expect(selectSpy).toHaveBeenCalledWith("test label")
    })
})