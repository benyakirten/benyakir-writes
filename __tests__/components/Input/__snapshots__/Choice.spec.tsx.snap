// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Choice component should change the background and text color based on the value prop 1`] = `
.c0 {
  position: relative;
  z-index: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  margin: 1rem;
  margin-left: 0;
  height: auto;
  width: 24%;
  border-radius: 2px;
  background-color: #343a40;
  color: #FDF3D8;
  font-size: 1.8rem;
  -webkit-transition: all 0.8s;
  transition: all 0.8s;
}

@media (max-width:922px) {
  .c0 {
    width: 45%;
  }
}

@media (max-width:576px) {
  .c0 {
    width: 75%;
  }
}

<div
  aria-checked={false}
  aria-labelledby="test label-label"
  checked={false}
  className="c0"
  data-cy="select-test-label"
  hidden={false}
  onClick={[Function]}
  role="checkbox"
  tabIndex={0}
>
  <span
    id="test label-label"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    test label
  </span>
</div>
`;

exports[`Choice component should change the background and text color based on the value prop 2`] = `
.c0 {
  position: relative;
  z-index: 0;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  margin: 1rem;
  margin-left: 0;
  height: auto;
  width: 24%;
  border-radius: 2px;
  background-color: #FDF3D8;
  color: #343a40;
  font-size: 1.8rem;
  -webkit-transition: all 0.8s;
  transition: all 0.8s;
}

@media (max-width:922px) {

}

@media (max-width:576px) {

}

@media (max-width:922px) {
  .c0 {
    width: 45%;
  }
}

@media (max-width:576px) {
  .c0 {
    width: 75%;
  }
}

<div
  aria-checked={true}
  aria-labelledby="test label-label"
  checked={true}
  className="c0"
  data-cy="select-test-label"
  hidden={false}
  onClick={[Function]}
  role="checkbox"
  tabIndex={0}
>
  <span
    id="test label-label"
    style={
      Object {
        "height": "auto",
      }
    }
  >
    test label
  </span>
</div>
`;
